File: C2/C2M2/CardGame/CardGame/Models/Player.cs
namespace CardGame.Models
{
    public abstract class Player : IComparable<Player>
    {
        public int Order { get; private set; }

        public string Name { get; protected set; }

        public CardGameApp CardGame { get; private set; }

        public int Point { get; private set; }

        public Hand Hand { get; private set; } = new();

        public void SetOrder(int order)
        {
            this.Order = order;
        }

        public abstract void NameHimself();

        public void SetCardGame(CardGameApp cardGame)
        {
            this.CardGame = cardGame;
        }

        public void GainPoint()
        {
            this.Point++;
        }

        public void SetHand(Hand hand)
        {
            Hand = hand;
        }

        public abstract Card Showdown();

        public int CompareTo(Player other)
        {
            return Point - other.Point;
        }
    }
}

File: C2/C2M2/c2m2h1-card-game-framework/game/showdown/Player.java
package game.showdown;

/**
 * @author - johnny850807@gmail.com (Waterball)
 */
public abstract class Player extends game.base.Player<Card> {
    protected int point;
    protected Showdown showdown;

    protected abstract Card takeTurn();

    public void setShowdown(Showdown showdown) {
        this.showdown = showdown;
    }

    public void gainPoint() {
        point++;
    }

    public int getPoint() {
        return point;
    }
}